cmd/scollector/collectors/haproxy_unix.go

```diff
@@ -0,0 +1,518 @@
+	var err error
+	const metric = "haproxy"
+	parse := func(v string) (int64, error) {
+		var i int64
+		if v != "" {
+			i, err = strconv.ParseInt(v, 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			return i, nil
+		}
+		return i, nil
+	}
+
+	req, err := http.NewRequest("GET", instance.URL, nil)
+	if err != nil {
+		return err
+	}
+	req.SetBasicAuth(instance.Username, instance.Password)
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	reader := csv.NewReader(resp.Body)
+	records, err := reader.ReadAll()
+	if err != nil {
+		return err
+	}
+	if len(records) < 2 {
+		return nil
+	}
+	for _, rec := range records[1:] {
+		// There is a trailing comma in haproxy's csv
+		if len(rec) != len(haproxyCSVMeta)+1 {
+			return fmt.Errorf("expected %v lines. got: %v",
+				len(haproxyCSVMeta)+1, len(rec))
+		}
+		hType := haproxyType[rec[32]]
+		pxname := rec[0]
+		svname := rec[1]
+		ts := opentsdb.TagSet{"pxname": pxname, "svname": svname, "tier": instance.Tier}
+		for i, field := range haproxyCSVMeta {
+			m := strings.Join([]string{metric, hType, field.Name}, ".")
+			switch i {
+			case 0, 1, 26, 27, 28, 31, 32, 56, 57:
+				continue
+			case 39, 40, 41, 42, 43, 44:
+				sp := strings.Split(field.Name, "_")
+				if len(sp) != 2 {
+					return fmt.Errorf("unexpected field name %v in hrsp", field.Name)
+				}
+				ts := ts.Copy().Merge(opentsdb.TagSet{"status_code": sp[1]})
+				m = strings.Join([]string{metric, hType, sp[0]}, ".")
+				v, err := parse(rec[i])
+				if err != nil {
+					return err
+				}
+				Add(md, m, v, ts, metadata.Counter, metadata.Response,
+					fmt.Sprintf("The number of http responses with a %v status code.", sp[1]))
+			case 17:
+				v, ok := haproxyStatus[rec[i]]
+				// Not distinging between MAINT and MAINT via...
+				if !ok {
+					v = 3
+				}
+				Add(md, m, v, ts, field.RateType, field.Unit, field.Desc)
+			case 36:
+				if rec[i] == "" {
+					continue
+				}
+				v, ok := haproxyCheckStatus[rec[i]]
+				if !ok {
+					return fmt.Errorf("unknown check status %v", rec[i])
+				}
+				Add(md, m, v, ts, field.RateType, field.Unit, field.Desc)
+			default:
+				v, err := parse(rec[i])
+				if err != nil {
+					return err
+				}
+				Add(md, m, v, ts, field.RateType, field.Unit, field.Desc)
+			}
+		}
+	}
+	return nil
+}
+
+type MetricMetaName struct {
+	Name string
+	MetricMeta
+}
+
+var haproxyType = map[string]string{
+	"0": "frontend",
+	"1": "backend",
+	"2": "server",
+	"3": "listen",
+}
+
+var haproxyCheckStatus = map[string]int{
+	"UNK":     0,
+	"INI":     1,
+	"SOCKERR": 2,
+	"L4OK":    3,
+	"L4TMOUT": 4,
+	"L4CON":   5,
+	"L6OK":    6,
+	"L6TOUT":  7,
+	"L6RSP":   8,
+	"L7OK":    9,
+	"L7OKC":   10,
+	"L7TOUT":  11,
+	"L7RSP":   12,
+	"L7STS":   13,
+}
+
+var haproxyStatus = map[string]int{
+	"UP":    0,
+	"DOWN":  1,
+	"NOLB":  2,
+	"MAINT": 3,
+}
+
+var haproxyCSVMeta = []MetricMetaName{
```

Comment:
maybe we could come up with a data structure that combines the bosun metadata with the haproxy field index metadata and drive all of the logic from data rather than the magic number switch thing.